### ISSUE TITLE
Project not working on dotnet core 3.0

### ISSUE THREAD

**Issue by Job79** [jchristn/WatsonTcp#84]
I installed the newest nuget package (4.1.12.2) and tried this project with dotnet core 3.0 (arch linux, kernel 5.7.4)

This is my code:
```cs
using System;
using WatsonTcp;

namespace WatsonTcpTest
{
    class Program
    {
        private const ushort Port = 42512;
        static void Main(string[] args)
        {
            using var server = new WatsonTcpServer("127.0.0.1", Port);
            server.ClientConnected += (s, client) => Console.WriteLine("Client connected");
            server.ClientDisconnected += (s, client) => Console.WriteLine("Client connected");
            server.MessageReceived += (s, message) => Console.WriteLine("Message received");
            Console.WriteLine("Starting server...");
            server.Start();

            using var client = new WatsonTcpClient("127.0.0.1", Port);
            client.MessageReceived += (o, s) => Console.WriteLine("Message received");
            Console.WriteLine("Connecting client...");
            client.Start();

            if (!client.Connected) return;
            else Console.WriteLine("Client connected");
            
            Console.WriteLine("Sending data...");
            client.Send("Test");
            Console.ReadLine();
        }
    }
}
```

Most of the time the client does not even connect to the server.
When it connects data is never received.
Am I doing something wrong?
Or does this project not work on linux? (I don't have a windows machine to test this code on)

**Comment by Job79** [comment]
Not having this issue with [SimpleTcp](https://github.com/jchristn/SimpleTcp)

**Comment by jchristn** [comment]
Hi @Job79 doesn't look like anything is wrong.  For grins, could you try putting a Thread.Sleep(1000) in between server.Start() and the client using statement?

**Comment by jchristn** [comment]
Actually try this and let me know.  I'm curious if the server/client are being disposed before execution of the 2nd line after.
```
using System;
using WatsonTcp;

namespace WatsonTcpTest
{
    class Program
    {
        private const ushort Port = 42512;
        static void Main(string[] args)
        {
            using (var server = new WatsonTcpServer("127.0.0.1", Port))
            {
                server.ClientConnected += (s, client) => Console.WriteLine("Client connected");
                server.ClientDisconnected += (s, client) => Console.WriteLine("Client connected");
                server.MessageReceived += (s, message) => Console.WriteLine("Message received");
                Console.WriteLine("Starting server...");
                server.Start();

                using (var client = new WatsonTcpClient("127.0.0.1", Port))
                {
                    client.MessageReceived += (o, s) => Console.WriteLine("Message received");
                    Console.WriteLine("Connecting client...");
                    client.Start();

                    if (!client.Connected) return;
                    else Console.WriteLine("Client connected");
                    
                    Console.WriteLine("Sending data...");
                    client.Send("Test");
                    Console.ReadLine();
                }
            }
        }
    }
}
```

**Comment by Job79** [comment]
This is the output I got:
```cs
Starting server...
Connecting client...
Client connected (client)
Sending data...
Client connected (server)
```
Maybe the server listener didn't start yet.
But putting a sleep before `Console.WriteLine("Sending data...");` doesn't seem to work.

**Comment by Job79** [comment]
> Actually try this and let me know. I'm curious if the server/client are being disposed before execution of the 2nd line after the using statement in your code since the code following it is not encapsulated in brackets.
> 
> ```
> using System;
> using WatsonTcp;
> 
> namespace WatsonTcpTest
> {
>     class Program
>     {
>         private const ushort Port = 42512;
>         static void Main(string[] args)
>         {
>             using (var server = new WatsonTcpServer("127.0.0.1", Port))
>             {
>                 server.ClientConnected += (s, client) => Console.WriteLine("Client connected");
>                 server.ClientDisconnected += (s, client) => Console.WriteLine("Client connected");
>                 server.MessageReceived += (s, message) => Console.WriteLine("Message received");
>                 Console.WriteLine("Starting server...");
>                 server.Start();
> 
>                 using (var client = new WatsonTcpClient("127.0.0.1", Port))
>                 {
>                     client.MessageReceived += (o, s) => Console.WriteLine("Message received");
>                     Console.WriteLine("Connecting client...");
>                     client.Start();
> 
>                     if (!client.Connected) return;
>                     else Console.WriteLine("Client connected");
>                     
>                     Console.WriteLine("Sending data...");
>                     client.Send("Test");
>                     Console.ReadLine();
>                 }
>             }
>         }
>     }
> }
> ```

Does this work on windows?

**Comment by jchristn** [comment]
Yes.

![image](https://user-images.githubusercontent.com/1421081/85929495-3e54a400-b86a-11ea-9b03-38aad0a68727.png)

**Comment by jchristn** [comment]
I also tried on .NET Core 3.0 (the above was from 3.1) and it worked.

![image](https://user-images.githubusercontent.com/1421081/85929513-5e846300-b86a-11ea-9293-4f33625c0adc.png)

**Comment by jchristn** [comment]
I'll get an Ubuntu image going and test this same code there too.  Let me know if it works on your end.  It should work fine, lots of people using it with Linux and .NET Core 3.0.

**Comment by Job79** [comment]
> I'll get an Ubuntu image going and test this same code there too. Let me know if it works on your end. It should work fine, lots of people using it with Linux and .NET Core 3.0.

Sure, tried it on my debian (arm) server. Same results there.
It is indeed a strange issue, normal sockets are working fine on linux.

**Comment by jchristn** [comment]
Check if it's using IPv6 instead of IPv4.  Add ```+ client.IpPort``` to the end of your client connected statement.  I'm testing now.

**Comment by Job79** [comment]
> Check if it's using IPv6 instead of IPv4. Add `+ client.IpPort` to the end of your client connected statement. I'm testing now.

It is using ipv4 ("127.0.0.1"), when replacing the addresses with ipv6 addresses same problem.

**Comment by jchristn** [comment]
Right, but I'm seeing an exception that leads me to believe somehow IPv6 is in the picture.
```
Unhandled exception. System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (111): Connection refused [::ffff:127.0.0.1]:42512
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(Exception source)
   at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)
   at System.Net.Sockets.TcpClient.EndConnect(IAsyncResult asyncResult)
   at WatsonTcp.WatsonTcpClient.Start()
   at WatsonTcpTest.Program.Main(String[] args) in c:\code\watson\WatsonTcpTest\WatsonTcpTest\Program.cs:line 24
```

**Comment by Job79** [comment]
> Right, but I'm seeing an exception that leads me to believe somehow IPv6 is in the picture.
> 
> ```
> Unhandled exception. System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (111): Connection refused [::ffff:127.0.0.1]:42512
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(Exception source)
>    at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)
>    at System.Net.Sockets.TcpClient.EndConnect(IAsyncResult asyncResult)
>    at WatsonTcp.WatsonTcpClient.Start()
>    at WatsonTcpTest.Program.Main(String[] args) in c:\code\watson\WatsonTcpTest\WatsonTcpTest\Program.cs:line 24
> ```
> 
> i.e.
> 
> ```
> Connection refused [::ffff:127.0.0.1]:42512
> ```

This can be fixed by putting a sleep between the client and the server.

**Comment by jchristn** [comment]
It's working.
```
joel@ubuntu:~/Desktop/Code/watsontcptest/WatsonTcpTest$ dotnet bin/Debug/netcoreapp3.0/WatsonTcpTest.dll
Starting server...
Connecting client...
Client connected
Sending data...
Client connected: 127.0.0.1:59166
Data sent!
```
Using your original port: 42512
```
joel@ubuntu:~/Desktop/Code/watsontcptest/WatsonTcpTest$ dotnet bin/Debug/netcoreapp3.0/WatsonTcpTest.dll
Starting server...
Connecting client...
Client connected
Sending data...
Client connected: 127.0.0.1:41416
Data sent!
```
Code:
```
using System;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using WatsonTcp;

namespace WatsonTcpTest
{
    class Program
    {
        private const ushort Port = 42512;
        static void Main(string[] args)
        {
            using (var server = new WatsonTcpServer("127.0.0.1", Port))
            {
                server.ClientConnected += (s, client) => Console.WriteLine("Client connected: " + client.IpPort);
                server.ClientDisconnected += (s, client) => Console.WriteLine("Client disconnected: " + client.IpPort);
                server.MessageReceived += (s, message) => Console.WriteLine("Message received: " + Encoding.UTF8.GetString(message.Data));
                Console.WriteLine("Starting server...");
                server.Start();

		Thread.Sleep(1000);
                using (var client = new WatsonTcpClient("127.0.0.1", Port))
                {
                    client.MessageReceived += (o, s) => Console.WriteLine("Message received");
                    Console.WriteLine("Connecting client...");
                    client.Start();

                    if (!client.Connected) return;
                    else Console.WriteLine("Client connected");

                    Console.WriteLine("Sending data...");
		    client.Send("Test");
		    Console.WriteLine("Data sent!");
                    Console.ReadLine();
                }
            }
        }
    }
}
```
dotnet version
```
joel@ubuntu:~/Desktop/Code/watsontcptest/WatsonTcpTest$ dotnet --version
3.1.301
```

**Comment by jchristn** [comment]
You said you're using Ubuntu on ARM?

**Comment by Job79** [comment]
> It's working.
> 
> ```
> joel@ubuntu:~/Desktop/Code/watsontcptest/WatsonTcpTest$ dotnet bin/Debug/netcoreapp3.0/WatsonTcpTest.dll
> Starting server...
> Connecting client...
> Client connected
> Sending data...
> Client connected: 127.0.0.1:59166
> Data sent!
> ```
> 
> Using your original port: 42512
> 
> ```
> joel@ubuntu:~/Desktop/Code/watsontcptest/WatsonTcpTest$ dotnet bin/Debug/netcoreapp3.0/WatsonTcpTest.dll
> Starting server...
> Connecting client...
> Client connected
> Sending data...
> Client connected: 127.0.0.1:41416
> Data sent!
> ```
> 
> Code:
> 
> ```
> using System;
> using System.Text;
> using System.Threading;
> using System.Threading.Tasks;
> using WatsonTcp;
> 
> namespace WatsonTcpTest
> {
>     class Program
>     {
>         private const ushort Port = 42512;
>         static void Main(string[] args)
>         {
>             using (var server = new WatsonTcpServer("127.0.0.1", Port))
>             {
>                 server.ClientConnected += (s, client) => Console.WriteLine("Client connected: " + client.IpPort);
>                 server.ClientDisconnected += (s, client) => Console.WriteLine("Client disconnected: " + client.IpPort);
>                 server.MessageReceived += (s, message) => Console.WriteLine("Message received: " + Encoding.UTF8.GetString(message.Data));
>                 Console.WriteLine("Starting server...");
>                 server.Start();
> 
> 		Thread.Sleep(1000);
>                 using (var client = new WatsonTcpClient("127.0.0.1", Port))
>                 {
>                     client.MessageReceived += (o, s) => Console.WriteLine("Message received");
>                     Console.WriteLine("Connecting client...");
>                     client.Start();
> 
>                     if (!client.Connected) return;
>                     else Console.WriteLine("Client connected");
> 
>                     Console.WriteLine("Sending data...");
> 		    client.Send("Test");
> 		    Console.WriteLine("Data sent!");
>                     Console.ReadLine();
>                 }
>             }
>         }
>     }
> }
> ```
> 
> dotnet version
> 
> ```
> joel@ubuntu:~/Desktop/Code/watsontcptest/WatsonTcpTest$ dotnet --version
> 3.1.301
> ```

Your output does not contain `Data Received`

> You said you're using Ubuntu on ARM?

Debian on arm, Arch on x86

**Comment by Job79** [comment]
![2020-06-27-214327_1920x1080_scrot](https://user-images.githubusercontent.com/38859657/85930831-c1541980-b8bf-11ea-8b44-c32785c4ccf3.png)
This while loop seems to never stop

**Comment by Job79** [comment]
Could it be that it is a problem with "\r\n" and "\n"?
That somewhere in the code "\r\n" is used and somewhere else Enviroment.Newline? (linux: "\n", windows: "\r\n")

**Comment by jchristn** [comment]
Good catch on it not including Message received.

I think it must be related to the newline separator.  Let me work on this and report back.  Thanks for bringing it to my attention!

**Comment by jchristn** [comment]
You @Job79 are the man.  Was able to fix it and push to NuGet - v4.1.12.4: 

Output:
![image](https://user-images.githubusercontent.com/1421081/85931454-613a8480-b879-11ea-8bc2-c001ad073729.png)

NuGet: https://www.nuget.org/packages/WatsonTcp/4.1.12.4
Commit: https://github.com/jchristn/WatsonTcp/commit/ea379fc2c42d325371e7f6bce00a3018cb20d8ea

I'll let you close omce you confirm.

I have a repo you can clone to test: https://github.com/jchristn/WatsonTcpTest

Thanks again!

**Comment by Job79** [comment]
Works great!
